### PR TITLE
[AI] Extract work with designs from production to separate module

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
- Make design cache calculation lazy, no need to call it during turn preparation explicitly
- Separate function to two parts: outer is to normalize arguments and shape return type and inner to do data aggregation staff,
- Remove unused get_design_cost
- Use a new approach to import from submodule: `from turn_state.design import ...`, this allows to handle circular imports,
- Annotate arguments and return types, update usages.

### CODE REVIEW

File: default/python/AI/ProductionAI.py [hunk 1]
```diff
@@ -13,178 +13,24 @@
...
-_design_cache = {}  # dict of tuples (rating,pid,designID,cost) sorted by rating and indexed by priority type
 
 _CHAT_DEBUG = False
 
 
-def find_best_designs_this_turn():
-    """Calculate the best designs for each ship class available at this turn."""
-    design_timer = AITimer('ShipDesigner')
-    design_timer.start('Updating cache for new turn')
-    ShipDesignAI.Cache.update_for_new_turn()
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py:42]: Need to check this if `ShipDesignAI.Cache.update_for_new_turn()` has temporal coupling with something else.

**Comment by Cjkjvfnby** [comment]
There is an example of how to deal with temporal coupling. 

Function `find_best_designs_this_turn` was called during the init of the turn.  
I have renamed it to `_fill_design_cache` and made all function who depends on it to call it explicitly.  
To avoid multiple calls of the `_fill_design_cache`  I have cached the result.  Also, I have made it return value instead of mutation of the module variable. This ensures that it will be called when needed.

Temporal coupling requires some shared state that needs to be changed in proper order.  Removing this state does a trick. 

This makes code easier. All work with the module is incapsulated inside it. You don't need to grep the whole project to check if some methods are called in proper order. Usually shared state is an implementation detail, and you don't want to make technical methods that init it as part of the public API of this module. 
Also, it makes code less error-prone. You don't break logic if you move some code in execution order. Sometimes correct order is done by accident rather than by design.

File: default/python/AI/ProductionAI.py (hunk 1 above)
Comment (Cjkjvfnby) [reply default/python/AI/ProductionAI.py:42]: Add `ShipDesignAI.Cache.update_for_new_turn` to turn init code.

**Comment by geoffthemedio** [comment]
is this done or still work in progress?

**Comment by Cjkjvfnby** [comment]
> is this done or still work in progress?

Yes, it's done

**Comment by geoffthemedio** [comment]
With this applied, but not in master, I get an error from an AI the turn after loading this save...

[nonetype_test.zip](https://github.com/freeorion/freeorion/files/5954769/nonetype_test.zip)

````
[10 Feb 00:05:10] AI_4: AI_Error: AI script error : FreeOrionAI.py:generateOrders():352  - Exception 'NoneType' object has no attribute 'productionCost' while trying to generate_production_orders
Traceback (most recent call last):
  File "C:/Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default\python\AI\FreeOrionAI.py", line 349, in generateOrders
    action()
  File "C:/Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default\python\AI\ProductionAI.py", line 1221, in generate_production_orders
    per_turn_cost = (float(best_design.productionCost(empire.empireID, loc)) / best_design.productionTime(empire.empireID, loc))
AttributeError: 'NoneType' object has no attribute 'productionCost'
Local context: {'local_drydocks': [], 'species_map': {'SP_CHATO': [1044]}, 'aistate': <AIstate.AIstate object at 0x0450C050>, 'wasted_pp': 6.394509315490723, 'pty': 6, 'bh_pilots': [], 't_ind': 0.0, 'building_id': 5949, 'planet_set': (198, 1044), 'total_available_troops': 0, 'red_pilots': [], 'bld_name': 'BLD_NEUTRONIUM_FORGE', 'block_str': '1 x ', 'neighbors': <freeOrionAIInterface.IntVec object at 0x0449EC80>, 'queued_locs': [], 'ship_number': 1, 'safety_margin_met': True, 'homeworld': P_1044<Invincible Alzirr I>, 'available_pp': {(198, 1044): 10.839592933654785}, 'capital_id': 1044, 'loop_count': 1, 'filtered_priorities': {8: 7.0, 9: 0.0, 10: 8.0, 6: 0.0}, 'production_queue': <freeOrionAIInterface.productionQueue object at 0x0450DB10>, 'built_camp': False, 'bh_popctrs': [], 'best_pilot_facilities': {'BLD_SHIPYARD_BASE': [1044], 'BLD_SHIPYARD_ORBITAL_DRYDOCK': [1044]}, 'priority_id': 9, 'avail_pp': 6.394509315490723, 'capital_queued_buildings': [], 'total_military_ships': 1, 'num_outpost_fleets': 0, 'element': <freeOrionAIInterface.productionQueueElement object at 0x04504B10>, 'can_build_camp': False, 'sys_id': 192, 'colony_ship_map': {}, 'possible_building_type_ids': ['BLD_EVACUATION', 'BLD_MILITARY_COMMAND', 'BLD_SUPER_TEST'], 'total_pp_spent': 4.4450836181640625, 'military_emergency': False, 'dd_sys_id': 1038, 'pop_disqualified': True, 'queued_troop_ships': 0, 'total_pp': 10.839592933654785, 'loc': 1044, 'queue_index': 0, 'spec_name': 'SP_CHATO', 'max_outpost_fleets': 1, 'pop_ctrs': [1044, 198], 'queued_clny_bld_locs': [], 'planets_with_wasted_pp': {(198, 1044)}, 'blue_pilots': [], 'queued_building_names': [], 'sorted_priorities': [(10, 8), (8, 7), (11, 5), (6, 0), (7, 0), (9, 0)], '_': 1.0, 'best_design': None, 'red_popctrs': [], 'verbose_camp': False, 'enrgy_shipyard_locs': {}, 'blue_popctrs': [], 'sType': freeOrionAIInterface.starType.red, 'pids': (198, 2694), 'planet': P_198<Zora I>, 'priority': 6, 'capital_system_id': 1038, 'bldg': 5949, 'max_dock_builds': 0, 'already_got_extractor': False, 'building_type': <freeOrionAIInterface.buildingType object at 0x0450D0E8>, 't_pop': 10.75, 'dequeue_list': [], 'acirema_systems': {}, 'type_id': 'BLD_SUPER_TEST', 'building_name': 'BLD_AUTO_HISTORY_ANALYSER', 'building_ratio': 0.35, 'possible_building_types': ['BLD_EVACUATION', 'BLD_MILITARY_COMMAND', 'BLD_SUPER_TEST'], 'total_colony_fleets': 0, 'max_gggs': 1, 'universe': <freeOrionAIInterface.universe object at 0x045049D0>, 'cur_drydoc_sys': {1038}, 'queued_xeno_lab_locs': [], 'making_colony_ship': True, 'building_expense': 0.0, 'top_pilot_systems': {}, 'total_outpost_fleets': 1, 'best_design_id': None, 'c_ind': 0.0, 'priority_facilities': ['BLD_SHIPYARD_ENRG_SOLAR', 'BLD_SHIPYARD_CON_GEOINT', 'BLD_SHIPYARD_AST_REF', 'BLD_SHIPYARD_ENRG_COMP'], 'empire': E_4<Solidarity>, 'this_priority': 8, 'score': 494.0, 'empire_species': {'SP_CHATO': [198, 1044]}, 'build_choices': None, 'shipyard_type': <freeOrionAIInterface.buildingType object at 0x0450DC00>, 'buildings_to_scrap': ('BLD_EVACUATION', 'BLD_GATEWAY_VOID'), 'claimed_stars': {freeOrionAIInterface.starType.blue: [1176], freeOrionAIInterface.starType.white: [1026], freeOrionAIInterface.starType.yellow: [840, 1812, 2334], freeOrionAIInterface.starType.orange: [192, 456, 636, 852, 1038, 1362, 1416], freeOrionAIInterface.starType.red: [822, 1374, 1620]}, 'queued_outpost_ships': 1, 'scaling_power': 1.0, 'top_score': 8, 'local_priorities': {8: 7.0, 9: 0.0, 10: 8.0, 6: 0.0}, 'systems_with_species': KeysView(<freeorion_tools._freeorion_tools.ReadOnlyDict object at 0x04480970>), 'table': <common.print_utils.Table object at 0x044807B0>, 'can_prioritize_troops': False, 'max_colony_fleets': 1, 'top_locs': [], 'num_colony_fleets': 0, 'pid': 1044, 'mil_build_choices': [(1.0, 1044, 6129, D_6129<SY Phoenix Mk. 1>)], 'queued_colony_ships': {}, 'start_set': {192, 282, 1038, 852, 1620, 1050}, 'capital_buildings': ['BLD_SHIPYARD_BASE', 'BLD_SHIPYARD_ORBITAL_DRYDOCK', 'BLD_CULTURE_ARCHIVES', 'BLD_IMPERIAL_PALACE', 'BLD_AUTO_HISTORY_ANALYSER'], 'this_role': 6, 'covered_drydoc_locs': {192, 282, 1038, 852, 1620, 1050}, 'total_troop_ships': 0, 'current_turn': 22, 'queued_shipyard_locs': [], 'avail_troop_fleet_ids': [], 'this_spec': 'SP_CHATO', 'all_troop_fleet_ids': [], 'dest_set': {192, 282, 1038, 852, 1620, 1050}, 'local_top_pilots': {}, 'priority_type': 11, 'all_military_fleet_ids': [3018], 'queued_sys': set(), 'allocated_pp': {(198, 1044): 4.4450836181640625}, 'colony_systems': {1044: 1038, 198: 192}, 'priority_choices': [8, 8, 8, 8, 8, 8, 8, 10, 10, 10, 10, 10, 10, 10, 10], 'queued_bld_locs': [], 'making_outpost_ship': False, 'best_ships': {8: [60, D_60<SD_COLONY_SHIP>, [1044]], 9: [120, D_120<SD_TROOP_SHIP>, [1044]], 10: [6129, D_6129<SY Phoenix Mk. 1>, [1044]], 6: [96, D_96<SD_SCOUT>, [1044]]}, 'building': <freeOrionAIInterface.building object at 0x0450DD18>, 'max_defense_portion': 0.0, 'system_colonies': {192: {'pids': [198]}, 1038: {'pids': [1044]}}, 'colony_build_choices': [1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044], 'c_pop': 2.1673295497894287, 'production_priorities': {6: 0, 7: 0, 8: 7, 9: 0, 10: 8, 11: 5}, 'colony_bldg_entries': []}
````

**Comment by geoffthemedio** [comment]
with the same save file, after loading and ending turn:
````
[03 Apr 11:49:09] AI_3: AI_Error: AI script error : FreeOrionAI.py:generateOrders():352  - Exception 'NoneType' object has no attribute 'productionCost' while trying to generate_production_orders
Traceback (most recent call last):
  File "C:/Users\g_top\Desktop\FOSDK11\FreeOrionAlt\default\python\AI\FreeOrionAI.py", line 349, in generateOrders
    action()
  File "C:/Users\g_top\Desktop\FOSDK11\FreeOrionAlt\default\python\AI\ProductionAI.py", line 1218, in generate_production_orders
    per_turn_cost = (float(best_design.productionCost(empire.empireID, loc)) / best_design.productionTime(empire.empireID, loc))
AttributeError: 'NoneType' object has no attribute 'productionCost'
Local context: {'total_outpost_fleets': 1, 'species_map': {'SP_CHATO': [1044]}, 'queued_clny_bld_locs': [], 'pid': 1044, 'spec_name': 'SP_CHATO', 'military_emergency': False, 'total_pp_spent': 4.4450836181640625, 'building_id': 5949, 'max_dock_builds': 0, 'bld_name': 'BLD_NEUTRONIUM_FORGE', 'colony_ship_map': {}, 'blue_popctrs': [], 'total_available_troops': 0, 'bh_pilots': [], 'possible_building_types': ['BLD_EVACUATION', 'BLD_MILITARY_COMMAND', 'BLD_SUPER_TEST'], 'queue_index': 0, 'dest_set': {192, 282, 1038, 852, 1620, 1050}, 'total_colony_fleets': 0, 'capital_queued_buildings': [], 'building': <freeOrionAIInterface.building object at 0x03F7ACA0>, 'c_pop': 2.1673295497894287, 'making_colony_ship': True, 'colony_bldg_entries': [], 'this_role': 6, 'max_colony_fleets': 1, 'top_pilot_systems': {}, 'pop_disqualified': True, 'safety_margin_met': True, 'queued_sys': set(), 'pop_ctrs': [1044, 198], 'dd_sys_id': 1038, 'available_pp': {(198, 1044): 10.839592933654785}, 'neighbors': <freeOrionAIInterface.IntVec object at 0x040D16B0>, 'local_drydocks': [], 'building_ratio': 0.35, 'block_str': '1 x ', 'best_ships': {8: [60, D_60<SD_COLONY_SHIP>, [1044]], 9: [120, D_120<SD_TROOP_SHIP>, [1044]], 10: [6129, D_6129<SY Phoenix Mk. 1>, [1044]], 6: [96, D_96<SD_SCOUT>, [1044]]}, 'capital_buildings': ['BLD_SHIPYARD_BASE', 'BLD_SHIPYARD_ORBITAL_DRYDOCK', 'BLD_CULTURE_ARCHIVES', 'BLD_IMPERIAL_PALACE', 'BLD_AUTO_HISTORY_ANALYSER'], 'type_id': 'BLD_SUPER_TEST', 'local_top_pilots': {}, 'all_military_fleet_ids': [3018], 'cur_drydoc_sys': {1038}, 'building_expense': 0.0, 'colony_build_choices': [1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044, 1044], 'red_pilots': [], 'sys_id': 192, '_': 1.0, 'priority_id': 9, 'total_troop_ships': 0, 'local_priorities': {8: 7.0, 9: 0.0, 10: 8.0, 6: 0.0}, 'system_colonies': {192: {'pids': [198]}, 1038: {'pids': [1044]}}, 'loc': 1044, 'queued_bld_locs': [], 'element': <freeOrionAIInterface.productionQueueElement object at 0x03F7A228>, 'already_got_extractor': False, 'building_name': 'BLD_AUTO_HISTORY_ANALYSER', 'enrgy_shipyard_locs': {}, 'acirema_systems': {}, 'priority_type': 11, 'homeworld': P_1044<Invincible Alzirr I>, 'total_pp': 10.839592933654785, 'ship_number': 1, 'start_set': {192, 282, 1038, 852, 1620, 1050}, 'bldg': 5949, 'bh_popctrs': [], 'avail_troop_fleet_ids': [], 'scaling_power': 1.0, 'colony_systems': {1044: 1038, 198: 192}, 'buildings_to_scrap': ('BLD_EVACUATION', 'BLD_GATEWAY_VOID'), 'c_ind': 0.0, 'priority_choices': [8, 8, 8, 8, 8, 8, 8, 10, 10, 10, 10, 10, 10, 10, 10], 'max_outpost_fleets': 1, 'queued_building_names': [], 'empire_species': {'SP_CHATO': [198, 1044]}, 'queued_colony_ships': {}, 'priority': 6, 'this_spec': 'SP_CHATO', 'queued_outpost_ships': 1, 'pty': 6, 'possible_building_type_ids': ['BLD_EVACUATION', 'BLD_MILITARY_COMMAND', 'BLD_SUPER_TEST'], 'top_locs': [], 'covered_drydoc_locs': {192, 282, 1038, 852, 1620, 1050}, 'best_pilot_facilities': {'BLD_SHIPYARD_BASE': [1044], 'BLD_SHIPYARD_ORBITAL_DRYDOCK': [1044]}, 'max_defense_portion': 0.0, 'loop_count': 1, 'production_priorities': {6: 0, 7: 0, 8: 7, 9: 0, 10: 8, 11: 5}, 'score': 494.0, 'planet_set': (198, 1044), 'blue_pilots': [], 'shipyard_type': <freeOrionAIInterface.buildingType object at 0x0404E8E0>, 'queued_shipyard_locs': [], 'empire': E_4<Solidarity>, 'build_choices': None, 'num_outpost_fleets': 0, 'universe': <freeOrionAIInterface.universe object at 0x03F7A250>, 'table': <common.print_utils.Table object at 0x046BA4B0>, 'dequeue_list': [], 'queued_xeno_lab_locs': [], 'capital_system_id': 1038, 'best_design': None, 'top_score': 8, 'systems_with_species': KeysView(<freeorion_tools._freeorion_tools.ReadOnlyDict object at 0x040C4950>), 'building_type': <freeOrionAIInterface.buildingType object at 0x0404EBB0>, 'built_camp': False, 'total_military_ships': 1, 't_ind': 0.0, 'pids': (198, 2694), 'filtered_priorities': {8: 7.0, 9: 0.0, 10: 8.0, 6: 0.0}, 'avail_pp': 6.394509315490723, 'can_build_camp': False, 'queued_locs': [], 'queued_troop_ships': 0, 'allocated_pp': {(198, 1044): 4.4450836181640625}, 'wasted_pp': 6.394509315490723, 'all_troop_fleet_ids': [], 'verbose_camp': False, 'best_design_id': None, 'mil_build_choices': [(1.0, 1044, 6129, D_6129<SY Phoenix Mk. 1>)], 'aistate': <AIstate.AIstate object at 0x040E5D50>, 'red_popctrs': [], 'current_turn': 22, 't_pop': 10.75, 'capital_id': 1044, 'this_priority': 8, 'planets_with_wasted_pp': {(198, 1044)}, 'sorted_priorities': [(10, 8), (8, 7), (11, 5), (6, 0), (7, 0), (9, 0)], 'planet': P_198<Zora I>, 'priority_facilities': ['BLD_SHIPYARD_ENRG_SOLAR', 'BLD_SHIPYARD_CON_GEOINT', 'BLD_SHIPYARD_AST_REF', 'BLD_SHIPYARD_ENRG_COMP'], 'can_prioritize_troops': False, 'num_colony_fleets': 0, 'sType': freeOrionAIInterface.starType.red, 'max_gggs': 1, 'making_outpost_ship': False, 'claimed_stars': {freeOrionAIInterface.starType.blue: [1176], freeOrionAIInterface.starType.white: [1026], freeOrionAIInterface.starType.yellow: [840, 1812, 2334], freeOrionAIInterface.starType.orange: [192, 456, 636, 852, 1038, 1362, 1416], freeOrionAIInterface.starType.red: [822, 1374, 1620]}, 'production_queue': <freeOrionAIInterface.productionQueue object at 0x0404EA48>}
````

**Comment by Cjkjvfnby** [comment]
> with the same save file, after loading and ending turn:

It was an issue with my code.

**Comment by Cjkjvfnby** [comment]
@Morlic-fo , do you have some time to review this code?

**Comment by geoffthemedio** [comment]
>  get an error from an AI

seems to be fixed

**Comment by Morlic-fo** [comment]
seems ok.

**Pull request merged**